[Actions][ServiceNow] Allow to close serviceNow incident when alert resolves

js-jankisalvi opened the pull request:
## Summary

Fixes: https://github.com/elastic/kibana/issues/170522

This PR allows to `close service now incident` when alert is `recovered`

SN connector form shows only `correlation_id` field as it is mandatory field to close an incident. 

![Screenshot 2023-11-27 at 11 52 36](https://github.com/elastic/kibana/assets/117571355/1d722153-f77a-484a-b17b-13489f9d7666)

**How to test:**
- Create a rule and select serviceNow ITSM action with Run when option as Recovered
- Verify that it closes an incident in SN when Alert is recovered


### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

## Release notes
Auto close ServiceNow incidents when alerts are resolved

js-jankisalvi commented on x-pack/plugins/alerting/common/disabled_action_groups.ts:
```diff
@@ -8,7 +8,7 @@
 import { RecoveredActionGroup } from './builtin_action_groups';
 
 const DisabledActionGroupsByActionType: Record<string, string[]> = {
-  [RecoveredActionGroup.id]: ['.jira', '.servicenow', '.resilient'],
```
To allow `Recovered` option

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.ts:
```diff
@@ -249,6 +255,81 @@ export const createExternalService: ServiceFactory = ({
+
+      checkInstance(res);
+
+      const foundIncident = useTableApi ? res.data.result : res.data.result[0];
```
I don't think we need this check here. We use the Table API (`tableApiIncidentUrl`) for this request. The results will always be an array of items. Example:

```
{
    "result": [
        {
            "sys_id": "12231",
            "short_description": "test itsm",
            "sys_created_on": "2023-10-27 13:23:46",
            "correlation_id": "5869"
        }
    ]
}
```

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.ts:
```diff
@@ -249,6 +255,81 @@ export const createExternalService: ServiceFactory = ({
+
+      return foundIncident;
+    } catch (error) {
+      throw createServiceError(error, 'Unable to find incidents');
```
nit: `Unable to get incident by correlation ID ${correlationId}`

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.ts:
```diff
@@ -249,6 +255,81 @@ export const createExternalService: ServiceFactory = ({
+      } else if (correlationId) {
+        incidentToBeClosed = await getIncidentByCorrelationId(correlationId);
+      } else {
+        throw new Error('Invalid correlationId or incidentId.');
```
In which scenario we will reach this code?

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.ts:
```diff
@@ -249,6 +255,81 @@ export const createExternalService: ServiceFactory = ({
+        throw new Error('Invalid correlationId or incidentId.');
+      }
+
+      const res = await request({
```
What do you think about using the `updateIncident` method instead?

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.test.ts:
```diff
@@ -805,6 +876,220 @@ describe('ServiceNow service', () => {
+    describe('import set table', () => {
+      test('it closes the incident correctly with incident id', async () => {
+        const spyOnCloseIncident = jest.spyOn(service, 'closeIncident');
+        spyOnCloseIncident.mockImplementation(closeIncidentMock);
```
Why do we mock the `closeIncident` function and not `requestMock`?

js-jankisalvi replied:
I tried `requestMock ` but it is not working as expected because `closeIncident` and `getIncident` has different response (one with result as array and one with result as object) which throw error.

cnasikas replied:
You can mock `requestMock ` as you like, with multiple responses for different calls, and only in your test (or your `describe` block).

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.test.ts:
```diff
@@ -805,6 +876,220 @@ describe('ServiceNow service', () => {
     });
   });
 
+  describe('closeIncident', () => {
```
Could you please add some tests where we check that we get the incident by correlation ID correctly (we call `requestMock` with the correct  URL etc)?

cnasikas replied:
Ok I just saw that you have them in `x-pack/plugins/stack_connectors/server/connector_types/servicenow_itsm/service.test.ts`. Should we move them here?

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/servicenow_itsm/service.test.ts:
I am a bit confused 🙂. Should not these tests be in `x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.ts`?

cnasikas replied:
If I remember correctly before we did not have the `lib` folder but at some point we refactored to have one service for all ServiceNow connectors. Probably there are leftovers. Putting them in the `lib` makes more sense to me because they can be used by other connectors ServiceNow connectors. Up to you.

cnasikas commented on x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/actions/connector_types/servicenow_itsm.ts:
```diff
@@ -716,6 +716,51 @@ export default function serviceNowITSMTest({ getService }: FtrProviderContext) {
         });
+
+        describe('closeIncident', () => {
+          it('should get choices', async () => {
```
nit: `should close the incident`

cnasikas commented on x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/actions/connector_types/servicenow_itsm.ts:
```diff
@@ -716,6 +716,51 @@ export default function serviceNowITSMTest({ getService }: FtrProviderContext) {
+        describe('closeIncident', () => {
+          it('should get choices', async () => {
+            // create an incident with correlation id
+            await supertest
```
As we do not call SN, is this call needed?

cnasikas reviewed (commented):
I tested and everything is working as expected 🚀. I managed to create an incident and close it when the alert recovers. I left some comments.

js-jankisalvi commented on x-pack/plugins/stack_connectors/public/connector_types/servicenow_itsm/servicenow_itsm.tsx:
```diff
@@ -64,5 +69,18 @@ export function getServiceNowITSMConnectorType(): ConnectorTypeModel<
+    defaultActionParams: {
+      subAction: 'pushToService',
+      subActionParams: {
+        incident: { correlation_id: DEFAULT_CORRELATION_ID, short_description: '' },
```
Turns out removing `short_description: '' `from `defaultActionParams` in `x-pack/plugins/stack_connectors/public/connector_types/servicenow_itsm/servicenow_itsm.tsx` doesn't throw any typescript error. 😄

js-jankisalvi commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/schema.ts:
```diff
@@ -115,6 +115,13 @@ export const ExecutorSubActionGetIncidentParamsSchema = schema.object({
 });
 
+export const ExecutorSubActionCloseIncidentParamsSchema = schema.object({
+  incident: schema.object({
```
same here. To have similar structure as `pushToService`

cnasikas replied:
Why do we need to have a similar structure when we close the incident?

js-jankisalvi replied:
just to be aware that correlation_id and externalId are part of incident. But I am kay to remove incident

cnasikas replied:
Ok, no makes sense 🙂.

js-jankisalvi commented on x-pack/plugins/stack_connectors/public/connector_types/servicenow_itsm/servicenow_itsm.tsx:
```diff
@@ -53,6 +57,7 @@ export function getServiceNowITSMConnectorType(): ConnectorTypeModel<
       if (
         actionParams.subActionParams &&
         actionParams.subActionParams.incident &&
+        actionParams.subAction !== 'closeIncident' &&
```
Good idea 👍

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/schema.ts:
```diff
@@ -115,6 +115,13 @@ export const ExecutorSubActionGetIncidentParamsSchema = schema.object({
+export const ExecutorSubActionCloseIncidentParamsSchema = schema.object({
+  incident: schema.object({
+    externalId: schema.nullable(schema.string()),
+    correlation_id: schema.nullable(schema.string({ defaultValue: DEFAULT_ALERTS_GROUPING_KEY })),
```
Ok, let's keep it for consistency.

js-jankisalvi commented on x-pack/plugins/stack_connectors/public/connector_types/servicenow_itsm/servicenow_itsm.test.tsx:
```diff
@@ -39,6 +39,18 @@ describe('servicenow action params validation', () => {
     });
   });
 
+  test(`${SERVICENOW_ITSM_CONNECTOR_TYPE_ID}: action params validation succeeds when no short_description provided and subAction is closeIncident`, async () => {
```
I added test in [d627b86](https://github.com/elastic/kibana/pull/171760/commits/d627b8623cf4936f2a1b49d2cfba8419b85e52d0), is this what you meant?

cnasikas commented on x-pack/plugins/stack_connectors/public/connector_types/servicenow_itsm/servicenow_itsm_params.test.tsx:
```diff
@@ -299,5 +287,58 @@ describe('ServiceNowITSMParamsFields renders', () => {
+
+      expect(editAction.mock.calls[0][1]).toEqual({
+        incident: { correlation_id: 'updated correlation id' },
+        comments: [],
```
We shouldn't have `comment` here when you are in `recoverd` mode, right?

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.ts:
```diff
@@ -249,6 +255,72 @@ export const createExternalService: ServiceFactory = ({
+
+      checkInstance(res);
+
+      if (res.status === 404) {
```
By default `axios` will throw an error for status codes > 300 ([`validateStatus](https://github.com/axios/axios)`). Did you notice a different behavior? Also because this is a `find` request I think `res.data.result` will be empty and ServiceNow will not return a `404` error when the search does match our criteria.

js-jankisalvi replied:
hmm.. 
Then I will put a check directly in closeIncident when `incidentToBeClosed` is null we log that we didn't find any incident.

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.ts:
```diff
@@ -249,6 +255,72 @@ export const createExternalService: ServiceFactory = ({
+      if (correlationId == null && incidentId == null) {
+        throw new Error('No correlationId or incidentId found.');
+      } else if (incidentId) {
+        incidentToBeClosed = await getIncident(incidentId);
```
This could throw a 404 error, right?

js-jankisalvi replied:
Updated in [2cdac29](https://github.com/elastic/kibana/pull/171760/commits/2cdac29775d8ca3b3989ceb667e623ec9a6b033e)

js-jankisalvi replied:
updated in https://github.com/elastic/kibana/pull/171760/commits/7088a167bdb1d3ab97f992bb0274cca8d57d489d

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.ts:
```diff
@@ -249,6 +255,72 @@ export const createExternalService: ServiceFactory = ({
+      }
+
+      if (incidentToBeClosed?.state === 7) {
+        logger.warn(`Incident is already closed.`);
```
I think we should enrich the log message to include more information. If you see the log without having any context it is difficult to understand what this message is about. What about `[ServiceNow][CloseIncident] Incident with correlation_id: ${correlationID} or incidentId: ${incidentId} is closed.`

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/service.ts:
```diff
@@ -249,6 +255,72 @@ export const createExternalService: ServiceFactory = ({
+        };
+      }
+
+      if (incidentToBeClosed === null || incidentToBeClosed?.status === 'error') {
```
nit: I think is better to return this kind of checks (Return Early Pattern) the earliest possible. Wdyt?

cnasikas reviewed (commented):
Thanks for addressing my feedback! Some comments 🙂.

cnasikas commented on x-pack/plugins/stack_connectors/server/connector_types/lib/servicenow/utils.ts:
```diff
@@ -42,13 +42,21 @@ const createErrorMessage = (errorResponse?: ServiceNowError): string => {
-  );
+  ) as AxiosError;
+
+  serviceError.code = error.code;
```
I just saw that you casted it 🙂 (`as AxiosError`). Probably this is why, right? Why not create doing `new AxiosError`? Did you encounter any issues?

js-jankisalvi replied:
yeah, we cannot create` new AxiosError() ` because it is not a constructor 😄 
I got this error ` TypeError: _axios.AxiosError is not a constructor`

js-jankisalvi replied:
fixed in https://github.com/elastic/kibana/pull/171760/commits/6c2d3daaf7dadb8f4513b7b1486ea2eb9b277b35

js-jankisalvi commented:
> 1. if you try to close the incident through the actions API by using the `externalId` you get an error. It seems that we need to also provide the `close_notes: 'Closed by Caller'`. In the ServiceNow UI it is a required field when you try to close an incident.
> 2. If the incident is already closed the warning message is not logged.

Fixed both bugs and added tests in https://github.com/elastic/kibana/pull/171760/commits/e3b43c7f365aa70cfbda908f1cf5624740431e73
